_posts/2017-01-23-cpp-strings.md

```diff
@@ -0,0 +1,150 @@
+---
+title: Modeling strings with libpmemobj C++ bindings
+author: RobDickinson
+layout: post
+---
+
+C++ developers using libpmemobj have multiple options for modeling strings,
+depending on the size of the strings and whether they are fixed or varying
+in length.
+
+For small fixed-length strings, use of `p<char[size]>` is efficient and easy.
+For large or variable-length strings, or for structs or arrays of such
+strings, then things get a little more complicated.
+
+#### Avoiding use of std::string
+
+For starters, never use `p<std::string>` or `persistent_ptr<std::string>`.
+These forms might seem a logical place to start, but these implementations
+will not be power-fail safe (at best) and might be unstable (at worst).
+This advice applies to using `p<>` or `persistent_ptr<>` wrappers around
+any complex types that perform their own memory management.
+Don't make this obvious mistake!
+
+#### Avoiding arrays of persistent pointers to single chars
+
+Don't model a variable-length string with `p<char> X[]`. This is power-fail
+safe, but wastes a lot of memory. Let's consider what's happening behind the
+scenes here. With this representation, each char in the variable-length
+string will be stored as one char in persistent memory, plus another 16 bytes
+for the persistent pointer to that char. So writing a single char results in
+16x more bytes written than necessary. Reading a single char results in 16x
+more bytes read than necessary, since reading every persistent char requires
+dereferencing a 16-byte persistent pointer. Obviously this is quite wasteful.
+
+#### Using fixed-length strings inside persistent structs
+
+While `p<char[size]>` is the obvious starting point for modeling fixed-size
```

Comment:
`p<>` does not implement the subscript `operator[]` and therefore p<char[]> will not work, as there is no `operator=` for arrays. This is one of those not apparent problems. We have thought about make such constructs work, but `p<>` has a simple API and we would like to keep it this way and not do template magic do push in some more functionality.